### ISSUE TITLE
External login redirect URIs get broken behind SSL-terminating proxies

### ISSUE THREAD

**Issue by markrendle** [aspnet/Security#1861]
So, I've got an app running in K8s behind the Nginx ingress controller. That's doing SSL termination for me so the ASP.NET Core app is running on `ASPNETCORE_URLS=http://+:80` (which comes from the `aspnetcore-runtime` Docker image btw) and has no `https` endpoint. This is fine.

I add Microsoft Account authentication. Microsoft Account insists on `https` redirect URLs, as it very well should.

Because Nginx is proxying to port 80 and ASP.NET Core is listening on port 80, `Context.Request.Scheme` gets set to `http`.

`AuthenticationHandler` uses `Request.Scheme` to construct the `redirect_uri` value, so it becomes `http://myapp.com/signin-microsoft` and that gets rejected by login.live.com.

There are two workarounds to this problem:

- Disable the SSL termination on the proxy and just send the HTTPS traffic straight through, which means you've got to configure Kestrel with a certificate and all sorts of ew, or
- Add the following snippet of middleware to my app, which is *horrible* but it's easy and it works:

```csharp
app.Use(async (context, next) =>
{
    // Override scheme to be https
    context.Request.Scheme = "https";
    await next();
});
```

Yeah, I wrote that code and it's in production right now.

Seems like it should be possible to override the `Request.Scheme` with an option, or something?

**Comment by Tratcher** [comment]
Have you looked at the reverse proxy [troubleshooting](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.1#troubleshoot) doc yet?

For containers we've seen a few reports of issues with IPv6 mapped IPv4 addresses confusing the [forwarder](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.1#configuration-for-an-ipv4-address-represented-as-an-ipv6-address). That's been fixed for 2.2.0-preview3 and there's a workaround in the doc.

**Comment by Eilon** [comment]
Closing for now. Please let us know if there is still a problem after reviewing the docs.
